### ISSUE TITLE
unneeded_field_pattern triggers in macro expansions

### ISSUE THREAD

**Issue by CobaltCause** [rust-lang/rust-clippy#11008]
### Summary

Clippy triggers `unneeded_field_pattern` on code inside a macro and suggests fixes that can't be applied because of that. Not totally sure if this is something clippy can do anything about, maybe the solution is adding the relevant `#[allow]` in the macro body.

### Lint Name

unneeded_field_pattern

### Reproducer

I tried this code:

```rust
#![warn(clippy::unneeded_field_pattern)]

use frunk::{hlist, hlist_pat}; // 0.4.2

enum Foo {
    Bar,
    Baz,
}

fn qux(foo: Foo) {
    let hlist_pat![a, _, b, c] = hlist![0_u8, foo, 1_u16, 2_u32];
    let _ = (a, b, c);
}
```

I saw this happen:

```
warning: you matched a field with a wildcard pattern, consider using `..` instead
  --> src/lib.rs:11:9
   |
11 |     let hlist_pat![a, _, b, c] = hlist![0_u8, foo, 1_u16, 2_u32];
   |         ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try with `HCons { tail: $crate::hlist_pat![$($tok)*], .. }`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unneeded_field_pattern
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![warn(clippy::unneeded_field_pattern)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this warning originates in the macro `$crate::hlist_pat` which comes from the expansion of the macro `hlist_pat` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see no correction.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```
